src/autoloads/characters.gd

```diff
@@ -155,34 +159,105 @@ func _process(delta: float) -> void:
 			positions[characterId] = _characterResources[characterId].getPosition()
-		rpc("_updateAllCharacterPositions", positions)
+		##
+		if len(serverSendQueue) > 0:
+			receiveCharacterDataServer(1, serverSendQueue)
+			serverSendQueue = []
+		## Broadcast all character data
+		#if len(broadcastDataQueue) > 0:
+			#print_debug(broadcastDataQueue)
+		rpc("_updateAllCharacterData", positions, broadcastDataQueue)
+		broadcastDataQueue = []
 	## If client
 	elif Connections.isClient():
 		if not get_tree().get_network_unique_id() in _characterResources:
 			return
 		## Send own character position to server
-		_sendMyCharacterPosToServer()
+		_sendMyCharacterDataToServer()
+	else:
+		assert(false, "Unreachable")
+
+## --Client functions
+
+func requestCharacterData() -> void:
+	rpc_id(1, "sendAllCharacterData")
 
 # puppet keyword means that when this function is used in an rpc call
 # 	it will only be run on client
-puppet func _updateAllCharacterPositions(positions: Dictionary) -> void:
+puppet func _updateAllCharacterData(positions: Dictionary, characterData: Array) -> void:
+	var myId: int = get_tree().get_network_unique_id()
 	## Loop through all characters
 	for characterId in positions:
 		# if this position is for this client's character
-		if characterId == get_tree().get_network_unique_id():
+		if characterId == myId:
 			# don't update its position
 			continue
 		## Set the position for the character
 		getCharacterResource(characterId).setPosition(positions[characterId])
-
-# --Server Functions--
+	## Decompose character data
+	if len(characterData) > 0:
+		print_debug(characterData)
+	for data in characterData:
+		if data["to"] == myId or data["to"] == -1:
+			receiveCharacterDataClient(data["id"], data["data"])
+
+func sendOwnCharacterData() -> void:
+	var id: int = get_tree().get_network_unique_id()
+	var characterRes: CharacterResource
+	characterRes = Characters.getCharacterResource(id)
+	var characterData: Dictionary = {}
+	characterData["outfit"] = characterRes.getOutfit()
+	characterData["colors"] = characterRes.getColors()
+	serverSendQueue.append(characterData)
+
+puppet func receiveCharacterDataClient(id: int, characterData: Dictionary) -> void:
+	var gameScene: Node = TransitionHandler.gameScene
+	gameScene.setCharacterData(id, characterData)
+
+## --Server Functions--
+
+master func sendAllCharacterData() -> void:
+	var characterRes: Dictionary = {}
+	characterRes = getCharacterResources()
+	for player in characterRes:
+		var characterData: Dictionary = {}
+		var outfit: Dictionary = characterRes[player].getOutfit()
+		var colors: Dictionary = characterRes[player].getColors()
+		if len(outfit) > 0:
+			characterData["outfit"] = outfit
+		if len(colors) > 0:
+			characterData["colors"] = colors
+		if len(characterData) > 0:
+			var senderId: int = get_tree().get_rpc_sender_id()
+			var dataSend: Dictionary = {"to": senderId, "id": player, "data": characterData}
+			broadcastDataQueue.append(dataSend)
```

Comment:
Could the data be sent here directly? The "to" parameter looks pretty weird

Reply:
Could the buffer be per-client, then? That could also be used for other types of messages.

Reply:
That's fair, the system can be improved later.